src/mongocxx/test/collection.cpp

```diff
@@ -2808,4 +2812,141 @@ TEST_CASE("Ensure that the WriteConcernError 'errInfo' object is propagated", "[
     REQUIRE(contains_err_info);
 }
 
+TEST_CASE("expose writeErrors[].errInfo", "[collection]") {
+
+    // A helper for checking that an error document is well-formed according to our requirements:
+    auto writeErrors_well_formed = [](const bsoncxx::document::view& reply_view) -> bool {
+        if (!reply_view["writeErrors"]) {
+            return false;
+        }
+
+        const auto& errdoc = reply_view["writeErrors"][0];
+
+        auto error_code = errdoc["code"].get_int32();
+
+        // The code should always be 121 (DocumentValidationFailure):
+        if (121 != error_code) {
+            std::ostringstream os;
+            os << "writeErrors expected to have code 121, but had " << error_code << " instead";
+            throw std::runtime_error(os.str());
+        }
+
+        // We require the "details" field be present:
+        if (!errdoc["errInfo"]["details"]) {
+            throw std::runtime_error("no \"details\" field in \"writeErrors\"");
+        }
+
+        return true;
+    };
+
+    // Set up our test environment:
+    instance::current();
+
+    auto client_opts = mongocxx::options::client();
+
+    // Listen to the insertion-failed event: we want to get a copy of the server's
+    // response so that we can compare it to the thrown exception later:
+    apm_opts.on_command_succeeded(
+        [&writeErrors_well_formed](const mongocxx::events::command_succeeded_event& ev) {
+            if(!writeErrors_well_formed(ev.reply()))
```

Comment:
This callback will be called on every command run on the client. That includes the "drop" command, the insert in the "insert a valid document" test.

To simplify this check, I suggest limiting this to "insert":
```suggestion
            if (0 != ev.command_name().compare("insert")) {
                return;
            }
            if(!writeErrors_well_formed(ev.reply()))
```